### ISSUE TITLE
Unnecessary Full-File Edits for Small Code Changes

### ISSUE THREAD

**Issue by kcui23** [cline/cline#583]
### Which API Provider are you using?

OpenRouter

### Which Model are you using?

Claude 3.5 Sonnet-beta

### What happened?

I’ve noticed a behavior that feels like a bug when working with large code files. When I need to make small changes—perhaps modifying only a few dozen lines—cline often invokes the `edit file` tool to update the entire file from the beginning, even when 99% of the file doesn’t need any changes. This significantly slows down the process.

While it’s known that truncating outputs in large files can be an issue, this behavior of editing the entire file seems unnecessary and inefficient when only small parts are being modified. It would be much more efficient if there were an option to generate code snippets directly in the conversation, which I can then manually copy and paste into the actual file.

**Expected Behavior**:

cline should provide an option to generate code snippets directly in the chat for minor edits, instead of always invoking the `edit file` tool to modify the entire file.

### Steps to reproduce

basically everytime working on large code files

**Comment by thunderbug1** [comment]
I also observed this, especially for larger files it can be a problem.

Maybe it would be an idea to give the llm an extra tool to apply patches to the file instead of a complete rewrite/diff cycle.
Patches could adhere to standard [patch file syntax ](https://www.oreilly.com/library/view/git-pocket-guide/9781449327507/ch11.html). I am not sure how good or bad llms are at writing patch files but since there should be quite a few patch files on the internet, they might have some in their training data.

Maybe this approach also better avoids the problem with the llm discarding parts of the file when rewriting it.

**Comment by thunderbug1** [comment]
I just checked, there are approaches to solve it [pr](https://github.com/cline/cline/pull/158) but it seems to still have some issues with result quallity.

**Comment by kcui23** [comment]
thanks for digging into this! It's pretty cool that there are already some attempts to tackle this problem out there.

**Comment by ccrvlh** [comment]
My previous experience with patch file syntax less than ideal - tricky to manage, tricky for the LLM's to get right, so I've been playing around with an alternative implementation, inspired by Aider's logic, added two tools: `Search And Replace` and `Insert Code Block` to try to work around this issue.

`Search And Replace` here: https://github.com/ccrvlh/cline/commit/c4b90017bb317144db1aa95c09893f3d5c79407a
`Insert Code Block` here: https://github.com/ccrvlh/cline/commit/a11a424df330179368265fd8e23e138ee1c2d066

Style/codebase is slightly different since I previously had separated the tools from the main `Cline` class into their own `tools.ts` file, however it should be simple enough to adapt for a PR if that's the case.

I've been testing the search and replace and works well, insert code block need a few tweaks on the identation and diffpreview part of things. here's a sneak peak of search and replace:

https://github.com/user-attachments/assets/dcd3006e-efec-4cf7-a02b-f9deba990558

**Comment by saoudrizwan** [comment]
I've experimented with structured output like editing certain lines or unified diff format like aider, but I've found the results aren't nearly as good as whole file outputs. The reason is these models are trained on significantly more whole code files than say diff patches, and there's been studies showing drops in performance whenever you force the models to output in a structured way ie diff or json.

OpenAI just released a new API featured called [predicted outputs](https://platform.openai.com/docs/guides/latency-optimization#use-predicted-outputs) that would fix this issue entirely. It essentially spits out the tokens its read before and only has to "think" about new tokens (the new edits to the file), making the edit faster and more reliable without the code truncation. The issue is that this API is very rudimentary and limited (doesn't allow images, only works with gpt4o, etc) so it doesn't fit as nicely into Cline as I'd like. I'm hoping Anthropic comes out with their version of this API soon (which they've pre-released in Zed, called [Fast Edit Mode](https://www.youtube.com/watch?v=6CCuyTMhZKw)), so I can add it to Cline ASAP and fix this problem.

Pls join the [discord](https://discord.gg/cline) for updates about this, and thank you for bearing with me in the meantime 🙏

**Comment by Kaddtechnologies** [comment]
why was this closed .this is still a huge problem and should be number 1 on the priority list

**Comment by mr-september** [comment]
Also please stop trying to re-direct discussion into discord. Discord is effectively not searchable, whether externally or internally.

**Comment by wesleyricarte** [comment]
I'm still facing the same problem. An option to that is strongly awaited.

**Comment by Javierg9n4** [comment]
Same here

**Comment by mmarras** [comment]
Same here
